frame/support/procedural/src/pallet/parse/call.rs

```diff
@@ -0,0 +1,237 @@
+		input.parse::<syn::Token![#]>()?;
+		let content;
+		syn::bracketed!(content in input);
+		content.parse::<keyword::pallet>()?;
+		content.parse::<syn::Token![::]>()?;
+
+		content.parse::<keyword::compact>()?;
+		Ok(ArgAttrIsCompact)
+	}
+}
+
+/// Check the syntax is `OriginFor<T>`
+pub fn check_dispatchable_first_arg_type(ty: &syn::Type) -> syn::Result<()> {
+	let expected = "expect `OriginFor<T>`";
+
+	pub struct CheckDispatchableFirstArg;
+	impl syn::parse::Parse for CheckDispatchableFirstArg {
+		fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+			input.parse::<keyword::OriginFor>()?;
+			input.parse::<syn::Token![<]>()?;
+			input.parse::<keyword::T>()?;
+			input.parse::<syn::Token![>]>()?;
+
+			Ok(Self)
+		}
+	}
+
+	syn::parse2::<CheckDispatchableFirstArg>(ty.to_token_stream())
+		.map_err(|e| {
+			let msg = format!("Invalid ty: {}", expected);
+			let mut err = syn::Error::new(ty.span(), msg);
+			err.combine(e);
+			err
+		})?;
+
+	Ok(())
+}
+
+impl CallDef {
+	pub fn try_from(
+		// Span needed for expansion
+		attr_span: proc_macro2::Span,
+		index: usize,
+		item: &mut syn::Item
+	) -> syn::Result<Self> {
+		let item = if let syn::Item::Impl(item) = item {
+			item
+		} else {
+			return Err(syn::Error::new(item.span(), "Invalid pallet::call, expect item impl"));
+		};
+
+		let mut instances = vec![];
+		instances.push(helper::check_impl_gen(&item.generics, item.impl_token.span())?);
+		instances.push(helper::check_pallet_struct_usage(&item.self_ty)?);
+
+		if let Some((_, _, for_)) = item.trait_ {
+			let msg = "Invalid pallet::call, expect no trait ident as in \
+				`impl<..> Pallet<..> { .. }`";
+			return Err(syn::Error::new(for_.span(), msg))
+		}
+
+		let mut methods = vec![];
+		for impl_item in &mut item.items {
+			if let syn::ImplItem::Method(method) = impl_item {
+				match method.sig.inputs.first() {
+					None => {
+						let msg = "Invalid pallet::call, must have at least origin arg";
+						return Err(syn::Error::new(method.sig.span(), msg));
+					},
+					Some(syn::FnArg::Receiver(_)) => {
+						let msg = "Invalid pallet::call, first argument must be a typed argument, \
+							e.g. `origin: OriginFor<T>`";
+						return Err(syn::Error::new(method.sig.span(), msg));
+					},
+					Some(syn::FnArg::Typed(arg)) => {
+						check_dispatchable_first_arg_type(&*arg.ty)?;
+					},
+				}
+
+				if let syn::ReturnType::Type(_, type_) = &method.sig.output {
+					syn::parse2::<keyword::DispatchResultWithPostInfo>(type_.to_token_stream())?;
+				} else {
+					let msg = "Invalid pallet::call, require return type \
+						DispatchResultWithPostInfo";
+					return Err(syn::Error::new(method.sig.span(), msg));
+				}
+
+				let mut call_var_attrs: Vec<FunctionAttr> =
+					helper::take_item_attrs(&mut method.attrs)?;
+
+				if call_var_attrs.len() != 1 {
+					let msg = if call_var_attrs.len() == 0 {
+						"Invalid pallet::call, require weight attribute i.e. `#[pallet::weight = $expr]`"
```

Comment:
Maybe not for this PR, but I think the mandatory weight attribute is pretty strong. Like given that we now have a base extrinsic weight, weight for any extrinsic will never litterally be zero...

What do you think @kianenigma

Reply:
I agree with it being mandatory. Writing a `weight == 0` is not too much of a hassle if your are testing things, and is a great reminder that you have to figure it out later.